### PR TITLE
📃 Property array support

### LINKED ISSUE #54: 📃 Property array support
> Refer [to the contributor docs](https://bitbra.in/pandora/#/contribute/adding-new-property) to learn how to generally add new properties to Pandora!

Similar to https://github.com/bitbrain/pandora/issues/95, we require a way to provide a "list" of entries - especially for recipes this is crucial.

Support any existing type, and via a `+` button additional elements can be added.

## Example use cases

- list of ingredients
- list of minerals to spawn
- factions that an NPC may be friendly with

### PR DESCRIPTION
**Pull request by mrfelipemartins**
## Description

Introducing support for Array Properties. Arrays can be of any currently supported property type except for Array. I tried to reuse existing property controls but a few tweaks was needed. New arrays are by default of type string, I believe we should have a popup for selecting the array type before it's created but I also think that this should be default for all properties that has settings so maybe this should be addressed in a separate PR.

## Addressed issues

- closes #54 

## Screenshots

image

image

### CODE REVIEW

File: addons/pandora/ui/editor/property_settings_editor/property_settings_editor.gd [hunk 1]
```diff
@@ -13,6 +13,15 @@ const EntityPicker = preload("res://addons/pandora/ui/components/entity_picker/e
 var _property:PandoraProperty
 var _default_settings:Dictionary
 
+var allowed_nested_types: Dictionary = {
```
Comment (bitbrain) [inline addons/pandora/ui/editor/property_settings_editor/property_settings_editor.gd]: This potentially can be improved by adding a new boolean method to `PandoraPropertyType` that determines if a type can be nested or not. All we then have to do is to add a new method to [type.gd](https://github.com/bitbrain/pandora/blob/godot-4.x/addons/pandora/model/type.gd) like this:
```gdscript
static func get_all_types() -> Array[PandoraPropertyType]
```
That can be used to find and lookup all types where the `allow_nesting()` may be true.

**Comment by bitbrain** [comment]
PR looks good already - I definitely have to try this out. In the meantime, could you also make sure to add some more unit tests for those new UI components you created?

File: addons/pandora/ui/editor/property_settings_editor/property_settings_editor.gd (hunk 1 above)
Comment (mrfelipemartins) [reply addons/pandora/ui/editor/property_settings_editor/property_settings_editor.gd]: I like that! Will update the PR

**Comment by mrfelipemartins** [comment]
Sure! I will create the tests. There are some improvements that I believe should be done, like ones we discussed on discord, but they should apply to all properties so I will create a separate issue to address them.

**Comment by bitbrain** [comment]
I have thoroughly tested this pull request and it looks mostly good to me, however, I found some flaws with it that we may need to address:

### Dubious error when adding an array of type `reference` and trying to add an item

```
Pandora error: value <null> is incompatible with type <RefCounted#-9223368694665487790>
```
The error disappeared after restarting Godot.

### Unable to configure the property type of array

One use case I have is that I have a bag with let's say 6 item slots. However, I only want to select entities of type `Item` and not get all the entities in the world suggested. To solve this, we could show as part of the array property settings the (proxied) settings of its type underneath it.
Screenshot 2023-09-15 175710

- When selecting the `reference` type, it displays all the config of `reference` underneath it. Those settings could function similar to what we already have: by default, it always uses the default settings coming with the `PandoraEntityType` - when the user then overrides stuff, the overrides could be stored inside the array settings under a special new key, e.g. `"type_settings": ...`
- whenever we add a new element to the array, we apply the settings to that entry
- changing a property value should apply this to all entries in the array (if any exist)

# Changing the array type does not give a warning

In case the array is not empty and I am changing the type, it just wipes the data. We may want to think about showing a popup to ask the user first. The same applies to other properties so finding a universal solution (as a separate PR) might be preferrable.

**Review by bitbrain — CHANGES_REQUESTED** [review]
as mentioned above.

**Review by bitbrain — APPROVED** [review]

**Comment by bitbrain** [comment]
LGTM - well done!

**Pull request merged**